File: android/app/src/main/java/com/noop/push/ServerVitalSelection.kt
package com.noop.push

/** Explicit Today vital mode: an absent selected-day server value never borrows local history. */
data class ServerVitalSelection(
    val value: Double?, val fromServer: Boolean, val day: String, val status: String?, val stale: Boolean,
    val sourceFeature: String? = null, val deviceId: String? = null, val algorithmVersion: String? = null,
) {
    enum class Metric { HRV, RESTING_HR, RESPIRATORY, SLEEP, CHARGE, STRAIN, SPO2, SKIN_TEMP }

    companion object {
        /** [overlay] must already be owner scoped and configuration/authentication qualified. */
        fun resolve(metric: Metric, serverEnabled: Boolean, selectedDay: String,
                    overlay: ServerScoreDayCache?, localValue: Double?): ServerVitalSelection {
            if (!serverEnabled) return ServerVitalSelection(localValue, false, selectedDay, null, false)
            // No overlay yet: keep showing locally scored values until the hosted scorer publishes.
            if (overlay == null || overlay.day != selectedDay)
                return ServerVitalSelection(localValue, false, selectedDay, null, false)
            val (value, featureKey) = when (metric) {
                Metric.HRV -> overlay.daily?.hrvRmssdMs to "hrv"
                Metric.RESTING_HR -> overlay.daily?.restingHrBpm?.toDouble() to "hrv"
                Metric.RESPIRATORY -> overlay.daily?.respRateBpm to "respiration"
                Metric.SLEEP -> overlay.daily?.sleepTotalMin to "sleep"
                Metric.CHARGE -> overlay.daily?.recovery to "hrv"
                Metric.STRAIN -> overlay.daily?.strain to "hrv"
                Metric.SPO2 -> overlay.daily?.spo2Pct to "hrv"
                Metric.SKIN_TEMP -> (overlay.daily?.skinTempC ?: overlay.daily?.skinTempDevC) to "hrv"
            }
            val feature = overlay.features[featureKey]
            val status = feature?.status ?: "unavailable"
            val available = status == "available" || status == "stale"
            // A published feature with this metric still null is not live for the card.
            // Keep the local number until the hosted kernel actually writes this key.
            if (!available || value == null) {
                return ServerVitalSelection(localValue, false, selectedDay, status, overlay.stale,
                    feature?.let { featureKey }, feature?.deviceId, feature?.algorithmVersion)
            }
            return ServerVitalSelection(value, true, selectedDay, status, overlay.stale || status == "stale",
                feature?.let { featureKey }, feature?.deviceId, feature?.algorithmVersion)
        }
    }
}

File: Packages/WhoopStore/Sources/WhoopStore/ServerVitalSelection.swift
import Foundation

/// Chooses one explicit mode for a Today vital. A missing server value never borrows local history.
/// The caller supplies an owner-scoped overlay only after configuration and authentication checks.
public struct ServerVitalSelection: Equatable {
    public enum Metric: CaseIterable { case hrv, restingHR, respiratory, sleep, charge, strain, spo2, skinTemp }
    public let value: Double?
    public let fromServer: Bool
    public let day: String
    public let status: String?
    public let stale: Bool
    public let sourceFeature: String?
    public let deviceId: String?
    public let algorithmVersion: String?

    public static func resolve(_ metric: Metric, serverEnabled: Bool, selectedDay: String,
                               overlay: ServerScoreDayCache?, localValue: Double?) -> Self {
        guard serverEnabled else {
            return Self(value: localValue, fromServer: false, day: selectedDay, status: nil, stale: false,
                        sourceFeature: nil, deviceId: nil, algorithmVersion: nil)
        }
        // No overlay yet: keep showing locally scored values until the hosted scorer publishes.
        guard let overlay, overlay.day == selectedDay else {
            return Self(value: localValue, fromServer: false, day: selectedDay, status: nil, stale: false,
                        sourceFeature: nil, deviceId: nil, algorithmVersion: nil)
        }
        let value: Double?
        let featureKey: String
        switch metric {
        case .hrv: value = overlay.daily?.hrvRmssdMs; featureKey = "hrv"
        case .restingHR: value = overlay.daily?.restingHrBpm.map(Double.init); featureKey = "hrv"
        case .respiratory: value = overlay.daily?.respRateBpm; featureKey = "respiration"
        case .sleep: value = overlay.daily?.sleepTotalMin; featureKey = "sleep"
        case .charge: value = overlay.daily?.recovery; featureKey = "hrv"
        case .strain: value = overlay.daily?.strain; featureKey = "hrv"
        case .spo2: value = overlay.daily?.spo2Pct; featureKey = "hrv"
        case .skinTemp: value = overlay.daily?.skinTempC ?? overlay.daily?.skinTempDevC; featureKey = "hrv"
        }
        let feature = overlay.features[featureKey]
        let status = feature?.status ?? "unavailable"
        let available = status == "available" || status == "stale"
        // A published feature with this metric still null is not live for the card.
        // Keep the local number until the hosted kernel actually writes this key.
        if !available || value == nil {
            return Self(value: localValue, fromServer: false, day: selectedDay, status: status, stale: overlay.stale,
                        sourceFeature: feature == nil ? nil : featureKey, deviceId: feature?.deviceId,
                        algorithmVersion: feature?.algorithmVersion)
        }
        return Self(value: value, fromServer: true, day: selectedDay,
                    status: status,
                    stale: overlay.stale || status == "stale", sourceFeature: feature == nil ? nil : featureKey,
                    deviceId: feature?.deviceId, algorithmVersion: feature?.algorithmVersion)
    }
}
